src/main/java/net/minecraftforge/common/data/JsonCodecProvider.java

```diff
@@ -0,0 +1,103 @@
+ * @param dynamicOps DynamicOps to encode values to jsons with using the provided Codec, e.g. {@link JsonOps.INSTANCE}.
+ * @param packType PackType specifying whether to generate entries in assets or data.
+ * @param directory String representing the directory to generate jsons in, e.g. "dimension" or "cheesemod/cheese".
+ * @param codec Codec to encode values to jsons with using the provided DynamicOps.
+ * @param entries Map of named entries to serialize to jsons. Paths for values are derived from the ResourceLocation's entryid:entrypath as specified above.
+ */
+public record JsonCodecProvider<T>(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper, String modid, DynamicOps<JsonElement> dynamicOps, PackType packType, String directory, Codec<T> codec, Map<ResourceLocation, T> entries) implements DataProvider
+{
+    private static final Logger LOGGER = LogManager.getLogger();
+    
+    public JsonCodecProvider
+    {
+        // Track generated data so other dataproviders can validate if needed.
+        final ResourceType resourceType = new ResourceType(packType, ".json", directory);
+        for (ResourceLocation id : entries.keySet())
+        {
+            existingFileHelper.trackGenerated(id, resourceType);
+        }
+    }
+    
+    /**
+     * {@return DatapackRegistryProvider that encodes using the registered loading codec for the provided registry key}
+     * Ensures the correct directory and codec are used.
+     * Only vanilla datapack registries enumerated in {@link RegistryAccess} and custom forge datapack registries can
+     * be generated this way.
+     * 
+     * @param <T> Registry element type, e.g. Biome
+     * @param dataGenerator DataGenerator provided by {@link GatherDataEvent}.
+     * @param modid namespace of the mod adding this DataProvider, for logging purposes.
+     * @param registryOps RegistryOps to encode values to json with.
+     * @param registryKey ResourceKey identifying the registry and its directory.
+     * @param entries Map of entries to encode and their ResourceLocations. Paths for values are derived from the ResourceLocation's entryid:entrypath.
+     */
+    public static <T> JsonCodecProvider<T> forDatapackRegistry(DataGenerator dataGenerator, ExistingFileHelper existingFileHelper, String modid, RegistryOps<JsonElement> registryOps, ResourceKey<Registry<T>> registryKey, Map<ResourceLocation, T> entries)
+    {
+        final ResourceLocation registryId = registryKey.location();
+        // Minecraft datapack registry folders are in data/json-namespace/registry-name/
+        // Non-vanilla registry folders are data/json-namespace/registry-namespace/registry-name/
+        final String registryFolder = registryId.getNamespace().equals("minecraft")
+            ? registryId.getPath()
+            : registryId.getNamespace() + "/" + registryId.getPath();
+        final Codec<T> codec = (Codec<T>)RegistryAccess.REGISTRIES.get(registryKey).codec();
```

Comment:
Formatting: space after cast.
```suggestion
        final Codec<T> codec = (Codec<T>) RegistryAccess.REGISTRIES.get(registryKey).codec();
```